Algorithm.Framework/Risk/MaximumSectorExposureRiskManagementModel.cs

```diff
@@ -0,0 +1,118 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using QuantConnect.Algorithm.Framework.Portfolio;
+using QuantConnect.Data.UniverseSelection;
+
+namespace QuantConnect.Algorithm.Framework.Risk
+{
+    /// <summary>
+    /// Provides an implementation of <see cref="IRiskManagementModel"/> that limits
+    /// the sector exposure to the specified percentage
+    /// </summary>
+    public class MaximumSectorExposureRiskManagementModel : IRiskManagementModel
+    {
+        private readonly decimal _maximumSectorExposure;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MaximumSectorExposureRiskManagementModel"/> class
+        /// </summary>
+        /// <param name="maximumSectorExposure">The maximum exposure for any sector, defaults to 20% sector exposure.</param>
+        public MaximumSectorExposureRiskManagementModel(
+            decimal maximumSectorExposure = 0.20m
+            )
+        {
+            _maximumSectorExposure = Math.Abs(maximumSectorExposure);
+        }
+
+        /// <summary>
+        /// Manages the algorithm's risk at each time step
+        /// </summary>
+        /// <param name="algorithm">The algorithm instance</param>
+        /// <param name="targets">The current portfolio targets to be assessed for risk</param>
+        public IEnumerable<IPortfolioTarget> ManageRisk(QCAlgorithmFramework algorithm, IPortfolioTarget[] targets)
+        {
+            var maximumSectorExposureValue = algorithm.Portfolio.TotalPortfolioValue * _maximumSectorExposure;
+
+            var targetCollection = new PortfolioTargetCollection();
+            targetCollection.AddRange(targets);
+
+            // Group the securities by their sector
+            var groupBySector = algorithm.Securities
```

Comment:
framework models should **not** loop over `algorithm.Securities` since it contains *all* securities that were ever added to the algorithm. Instead you should use `OnSecuritiesChanged` to keep a set of the current securities. You can check out `ConstantAlphaModel.OnSecuritiesChanged' for how he keeps track of current in-universe securities (we should definitely add a helper function for this). Here's the quick way of doing it:
``` csharp
NotifiedSecurityChanges.UpdateCollection(_securities, changes);
```

there's also an `UpdateDictionary` version of the method.